### PR TITLE
[FIX] mail: send is_main with the attachments

### PR DESCRIPTION
**Pull request by phenix-factory**
task-2783076

### CODE REVIEW

File: addons/mail/controllers/discuss.py [hunk 1]
```diff
@@ -551,6 +551,11 @@ def audio_worklet_processor(self):
             ]
         )
 
+    @http.route('/mail/thread/data', methods=['POST'], type='json', auth='user')
+    def mail_thread_attachment(self, thread_model, thread_id, **kwargs):
```
Comment (seb-odoo) [inline addons/mail/controllers/discuss.py:555]: forgot to rename the method following change

Comment (seb-odoo) [reply addons/mail/controllers/discuss.py:555]: also need to put the route in the same position in the file as in master

Comment (seb-odoo) [reply addons/mail/controllers/discuss.py:555]: also should have the same signature and return value

Comment (phenix-factory) [reply addons/mail/controllers/discuss.py]: Done !

Comment (seb-odoo) [reply addons/mail/controllers/discuss.py]: The route still doesn't take the request_list param and attachments return key.

Comment (phenix-factory) [reply addons/mail/controllers/discuss.py]: Yes, because it trigger a "var declared by unused" error on Runbot. I removed it because it's not used and didn't want to use it just because.

Comment (seb-odoo) [reply addons/mail/controllers/discuss.py]: But it needs to be used in a condition same as in master, we don't want the code to be any different.

File: addons/mail/static/src/models/thread/thread.js [hunk 2]
```diff
@@ -708,24 +708,37 @@ function factory(dependencies) {
         }
 
         /**
-         * Fetch attachments linked to a record. Useful for populating the store
-         * with these attachments, which are used by attachment box in the chatter.
-         */
-        async fetchAttachments() {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:714]: Keep the method in stable, just make it call the other one with the correct params.

File: addons/mail/static/src/models/thread/thread.js [hunk 3]
```diff
@@ -1075,7 +1088,7 @@ function factory(dependencies) {
             }
             this.loadNewMessages();
             this.update({ isLoadingAttachments: true });
```
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:1090]: This (and the one below) should probably be removed from here, as it is now done in the new method.

File: addons/mail/static/src/models/thread/thread.js [hunk 4]
```diff
@@ -707,25 +707,46 @@ function factory(dependencies) {
...
+            }, { shadow: true });
+            const values = {};
+            if (attachmentsData) {
+                Object.assign(values, {
+                    areAttachmentsLoaded: true,
+                    isLoadingAttachments: false,
+                    originThreadAttachments: insertAndReplace(attachmentsData),
+                });
+            }
+            this.update(values);
```
Comment (alexkuhn) [inline addons/mail/static/src/models/thread/thread.js:749]: Same remark as in the 14.0 PR, this looks like backport code. So could be isolated in its own commit, and then follow with commit that focuses precisely on fixing `is_main`,

Comment (phenix-factory) [reply addons/mail/static/src/models/thread/thread.js:749]: This one could be a better candidate for this, since attachment_format already exist in 15.0.
But then again, the whole code is a backport from master.

**Comment by seb-odoo** [comment]
Don't forget to remove the `is_main` from message format (now duplicate with the one in attachment).

File: addons/mail/models/ir_attachment.py [hunk 5]
```diff
@@ -58,6 +58,9 @@ def _attachment_format(self, commands=False):
                 'name': attachment.name,
                 'mimetype': 'application/octet-stream' if safari and attachment.mimetype and 'video' in attachment.mimetype else attachment.mimetype,
             }
+            if attachment.res_id and issubclass(self.pool[attachment.res_model], self.pool['mail.thread']):
+                main_attachment = self.env[attachment.res_model].sudo().browse(attachment.res_id).message_main_attachment_id
+                res['is_main'] = attachment.id == main_attachment.id
```
Comment (seb-odoo) [inline addons/mail/models/ir_attachment.py:63]: should not be comparing ids but directly records

File: addons/mail/static/tests/helpers/mock_server.js [hunk 6]
```diff
@@ -2114,4 +2118,42 @@ MockServer.include({
...
     },
+    /**
+     * Simulates `_attachment_format` on `ir.attachment`.
+     *
+     * @private
+     * @param {string} res_model
+     * @param {string} domain
+     * @returns {Object}
+     */
+    _mockIrAttachment_attachmentFormat(ids, commands = false) {
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:2129]: not positioned at the correct place in the file

File: addons/mail/static/tests/helpers/mock_server.js [hunk 7]
```diff
@@ -2114,4 +2118,42 @@ MockServer.include({
...
+            }
+            return res;
+        });
+    },
+    _mockRouteMailThreadAttachments(thread_model, thread_id) {
+        const attachments = this._mockSearchRead('ir.attachment', [
+            [['res_id', '=', thread_id], ['res_model', '=', thread_model]],
+        ], {}); // order not done for simplicity
+        return { attachments: this._mockIrAttachment_attachmentFormat(attachments.map(attachment => attachment.id), true) };
+    }
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:2158]: should be mock of `_mockRouteMailThreadData`

File: addons/mail/static/src/models/thread/thread.js [hunk 8]
```diff
@@ -707,25 +707,46 @@ function factory(dependencies) {
...
+                attachments: attachmentsData,
+            } = await this.env.services.rpc({
+                route: '/mail/thread/data',
+                params: {
+                    request_list: [...requestSet],
+                    thread_id: this.id,
+                    thread_model: this.model,
+                },
+            }, { shadow: true });
+            const values = {};
```
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:741]: missing exists quick return

File: addons/mail/static/tests/helpers/mock_server.js [hunk 9]
```diff
@@ -2114,4 +2149,23 @@ MockServer.include({
...
     },
+    /**
+     * Simulates the `/mail/thread/data` route.
+     *
+     * @param {string} thread_model
+     * @param {integer} thread_mid
+     * @param {string[]} request_list
+     * @returns {Object}
+     */
+    async _mockRouteMailThreadData(thread_model, thread_id, request_list) {
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:2160]: wrong place in the file, should be alphabetical with the others

File: addons/mail/static/tests/helpers/mock_server.js [hunk 10]
```diff
@@ -2114,4 +2149,23 @@ MockServer.include({
             starred_counter: this._getRecords('mail.message', [['starred_partner_ids', 'in', user.partner_id]]).length,
         };
     },
+    /**
+     * Simulates the `/mail/thread/data` route.
+     *
+     * @param {string} thread_model
+     * @param {integer} thread_mid
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:2156]: typo in param name (also in master, but it's fine to fix it here)

File: addons/mail/static/tests/helpers/mock_server.js [hunk 11]
```diff
@@ -310,6 +313,38 @@ MockServer.include({
...
 
+    /**
+     * Simulates `_attachment_format` on `ir.attachment`.
+     *
+     * @private
+     * @param {string} res_model
+     * @param {string} domain
+     * @returns {Object}
+     */
+    _mockIrAttachment_attachmentFormat(ids, commands = false) {
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:324]: wrong place, should be in mocked method category, not mocked routes

File: addons/mail/static/tests/helpers/mock_server.js [hunk 12]
```diff
@@ -135,6 +135,9 @@ MockServer.include({
             const { channel_id, min_id, max_id, limit } = args;
             return this._mockRouteMailChannelMessages(channel_id, max_id, min_id, limit);
         }
+        if (route === '/mail/thread/data') {
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:138]: wrong place, should be above mail/thread/messages

**Review by seb-odoo — APPROVED** [review]
@robodoo r+
